docs_src/design/adr/core/0026-Core-Data-Cache.md

```diff
@@ -0,0 +1,85 @@
+# Core Data Cache
+
+## Status
+
+**Draft**
+
+proposed: 4/4/22
+
+## Context
+There is a desire to have core data serve more as a data cache - that is, keeping only the "latest" readings.  There is also a desire to allow for queries of core data to serve up the "latest" readings.  This ADR defines new operations and APIs to satisfy these needs.
+
+### Queries to return "Latest"
+
+Core data will add the following new operations (and externally visible APIs) to provide the means to get the `latest` readings.
+
+- Query for the latest N readings by device name
+    - API: *GET* /reading/device/name/{name}/latest/{N}
+    - returns the last N readings for the specified device
+    - N must be > 0
+- Query for the latest N readings by device name and device resource
+    - API: *GET* /reading/device/name/{name}/resourceName/{resourceName}/latest/{N}
+    - returns the latest N readings for the specified device and specified device resource
+    - N must be > 0
+
+### Delete operations (to keep the number of readings to the "latest" or a database size limit)
+
+Core data will add the following new delete operations.  Note that only one will be exposed externally via API.  The other will be used internally by the service (when configured to use it).
+
+- Delete all but N readings (aka - the **Eaton Delete**; clean readings but preserve the latest readings)
+    - API:  *DELETE* /reading/keep/{N}
+    - if there are fewer than (or equal to) N readings in core data, then all readings would be kept and no readings would be removed
+    - if there are more than N readings in core data, then delete the oldest readings, but keep the youngest N readings
+    - N must be > 0 (N = 1 means keep only the latest reading)
+    - this delete option will be triggered via API either:
+        - manually by the adopter
+        - or by the scheduling service (the adopter would set up a new IntervalAction with the scheduling service)
+
+- Delete readings to keep the database at a certain size (aka - the **Tony Delete**; clean readings to preserve the number of readings in the database)
+    - API:  none.  This operation is not visible via REST API
+    - this operation, when called upon internally by core data, will remove readings (from oldest to newest) to keep the core data database within a specified size
+    - a configuration setting (`CacheDelete` in configuration) will determine how readings will be removed by this operation:
+        - option A (`CacheDelete="OnEach"`): remove one reading (the oldest reading) each time a new reading gets added
+            - this operation would be called as part of the core data add event/reading operations
+            - this option causes core data to operate like a real cache - keeping only the latest reading
+        - option B (`CacheDelete="OnCount"`): remove a specified number (`DeleteSize` in configuration) of readings when the number of readings exceeds a specified limit (`CacheSize` in configuration)
```

Comment:
I don't think we need two options here.  If the reading number reach the `CacheSize` and `DeleteSize=1`, it will work like `CacheDelete="OnEach"` anyway.
Also, `CacheDelete="OnCount"`, `CacheSize=1` and `DeleteSize=1` will work the same as `CacheDelete="OnEach"`